docs/design/proposals/TopologySupport.md

```diff
@@ -0,0 +1,232 @@
+
+## Ceph-CSI topology support design
+
+Supporting topology aware provisioning in Ceph-CSI reduces to solving the following problems as a result of the above discussion,
+
+### Determining node domain by the CSI nodeplugin
+
+Currently the Ceph-CSI nodeplugins have no information regarding the domains they belong to. The `node_id`, in `NodeGetInfo` response, itself is picked up from the pod manifest [7]. Kubernetes maintains failure-domains[8], but these cannot be passed in via the pod manifest as yet [13], as these are not supported by the downward APIs[14] (unlike the `node-id` for example).
+
+The CSI nodeplugins also do not have any specific domain knowledge to present themselves, and would hence rely on the domain information of the cluster that it runs on.
+
+The proposal is to, feed CSI nodeplugins a list of labels that it can read from the COs to determine its domain and return the same in it's response to a `NodeGetInfo` request.
+
+For example in the case of the CO being kubernetes the following maybe be passed as an ordered list of domain labels to advertise. The order of labels ensures hierarchical domain relationships. The nodeplugin would read the values of these labels, via kubernetes client API, on the node where it is running and respond with the same in the `NodeGetInfo` request.
+
+```
+# ceph csi ... --orchestrator="kubernetes" --domain-labels="failure-domain.beta.kubernetes.io/region;failure-domain.beta.kubernetes.io/zone;failure-domain.mycluster.io/rack" ...
+```
+
+**NOTE:** Current proposal is to support only kubernetes as the orchestrator
+
+### StorageClass changes to denote pools with domain affinity
+
+As noted earlier, creation of pools that have a domain affinity, is out of scope for the CSI plugins. However, when presented with a set of pools that have the said property, a `CreateVolume` request with `accessibility_requirements` specified, needs to choose the right pool to create the image in (in the case of RBD), or to redirect file's data to (in the case of CephFS).
+
+The proposal is to, add the specified pools and their domain affinity in the StorageClass (or related constructs in non-kubernetes COs), for the controller to choose from. The changes to the StorageClass is detailed in this issue [15].
+
+Further, the CSI journal that is maintained as an OMap within a **single** pool needs to continue even when the volumes are allocated from different pools. As, we need a single pool that can hold information about the various CSI volumes to maintain idempotent responses to various CSI RPCs.
+
+The proposal hence is to also include a `pool` parameter in the StorageClass for RBD based volumes, where the CSI OMap journal would be maintained. This can be one of the subset of pools in the list of domain affinity based pools, or a more highly available cross-domain pool.
```

Comment:
Are you only allowed to run a single (active) controller or will this new topology aware provisioning allow you to run an active controller per topology domain? If the former, it seems like you could effectively isolate the domains from each other and therefore you wouldn't need a centralized "global" inter-domain pool.

Reply:
We can run multiple controllers, but routing of requests to these controllers would be at random by the CO and not follow any domain rules. Every controller instance that advertises the same driver name can get the provision request.

The alternate is to split these into their own StorageClass, and hence use the domain pool to maintain the journal, but that would defeat the transparent single StorageClass for users to use.

NOTE: The default pool to store the CSI journal need not necessarily be inter-zone, at least in this proposal, the user can choose one of the domain pools as the journal store, instead of having to create an inter-zone pool for just the journal needs (if the said journal pool is in accessible though, then provisioning would halt, but not stage/publish of other domain volumes, as the image OMap would reside on the same pool as the image).

Reply:
Under AWZ, Azure, and GCP, are all PVs globally namespaced even if they are tied to a specific zone?

Reply:
> Under AWZ, Azure, and GCP, are all PVs globally namespaced even if they are tied to a specific zone?

No, in [AWS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/1369deccddef21d41b643e3fea667a17e2080014/pkg/driver/controller.go#L149) and [Azure](https://github.com/kubernetes-sigs/azuredisk-csi-driver/blob/faa49d43722b10bd4023712d072a289afe9afddd/pkg/azuredisk/controllerserver.go#L159) a PV is namespaced to a zone, in [GCP](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/4343cc3d577a37b10a94551698fc3c77a39fd976/pkg/gce-pd-csi-driver/controller.go#L130-L131) it can pick 2 zones if replication is enabled.